### PR TITLE
Fixes #3241 Improve logging for 2.0 replication functional debugging

### LINKED ISSUE #3241: Improve logging for 2.0 replication functional debugging
- [ ] Switch back to WS/WS+/WS++ for logging done by the go-blip library (via callback)
- [x] Use "Sync/Sync+" for SG contextual logging in blip_sync
- [x] Use "SyncMsg"/"SyncMsg+" in the same way "HTTP" and "HTTP+" are used for 1.x REST calls.  
  - [x] Should log "SyncMsg" once at the start of each handler.  SyncMsg should include at least the same information that was available for the corresponding HTTP request in 1.x.  Parameters should not be logged if they aren't present
     - getCheckpoint (all parameters)
     - setCheckpoint (all parameters)
     - subChanges (all parameters)
     - changes (length of array)
     - proposeChanges (length of array)
     - rev (all parameters except history)
     - getAttachment (all)
     - proveAttachment (all)
  - [x] Should log "SyncMsg+" in a defer when leaving each handler (note that async handlers like subChanges would need to log when their goroutine exits).  Should include status code similar to HTTP+
  - [x] Timing information on SyncMsg+ similar to HTTP+
  - [x] Numbering on the SyncMsg logging (like "#1"), so that we can correlate SyncMsg with their corresponding SyncMsg+
  - [x] Reduce blip context to 32 bit number

### PR DESCRIPTION
**Pull request by tleyden**
Fixed #3241 

This code in particular might need some explaining/justification:

```
+type SGLogger interface {
+	LogTo(key string, format string, args ...interface{})
+}
```

and it's use in:

```
// Helper for handling BLIP subChanges requests.  Supports Stringer() interface to log aspects of the request.
type subChanges struct {
	rq                    *blip.Message       // The underlying BLIP message
	logger                base.SGLogger       // A logger object which might encompass more state (eg, blipContext id)
	sinceZeroValueCreator SequenceIDGenerator // A sequence generator for creating zero'd since values
}

```

The idea was to refactor out the `subChanges` code from the handler so that `subChanges` could easily dump it's "state" as a `fmt.Stringer()` -- where the state here is the parameters in the `*blip.Request`.  

However, there was some code in particular that needed a logger:

```
func (s *subChanges) since() db.SequenceID {

	// Depending on the db sequence type, use correct zero sequence for since value
	sinceSequenceId := s.sinceZeroValueCreator()

	if sinceStr, found := s.rq.Properties["since"]; found {
		var err error
		if sinceSequenceId, err = db.ParseSequenceIDFromJSON([]byte(sinceStr)); err != nil {
			s.logger.LogTo("Sync", "%s: Invalid sequence ID in 'since': %s", s.rq, sinceStr)
			sinceSequenceId = db.SequenceID{}
		}
	}

	return sinceSequenceId

}
```

and it felt "right" to move the `since()` method code out of the handler and into `subChanges`, since `subChanges` could be tested in isolation (and possibly used elsewhere).   Since the logger in the handler was being wrapped to emit the blip context id, it wasn't possible to just call `base.LogTo()` from `subChanges`, and so that's why a handler is being passed in.

I only implemented the minimal log related messages in the `SGLogger` interface, but if we want to go this route then maybe it would make sense to flesh out that interface a bit more.

### CODE REVIEW

**Comment by tleyden** [comment]
Integration test:

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/331/

File: rest/blip_sync_messages.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+func (s *subChanges) since() db.SequenceID {
+
+	// Depending on the db sequence type, use correct zero sequence for since value
+	sinceSequenceId := s.sinceZeroValueCreator()
+
+	if sinceStr, found := s.rq.Properties["since"]; found {
+		var err error
+		if sinceSequenceId, err = db.ParseSequenceIDFromJSON([]byte(sinceStr)); err != nil {
+			s.logger.LogTo("Sync", "%s: Invalid sequence ID in 'since': %s", s.rq, sinceStr)
```
Comment (adamcfraser) [inline rest/blip_sync_messages.go]: The 1.x replicator returns an error for an invalid since value.  This looks like it's doing INFO level logging then silently resetting the since value to zero.  I think this should be returning an error in the subChanges response, shouldn't it?

File: rest/blip_sync_messages.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+	logger                base.SGLogger       // A logger object which might encompass more state (eg, blipContext id)
+	sinceZeroValueCreator SequenceIDGenerator // A sequence generator for creating zero'd since values
+}
+
+// Create a new subChanges helper
+func newSubChanges(rq *blip.Message, logger base.SGLogger, sinceZeroValueCreator SequenceIDGenerator) *subChanges {
+	return &subChanges{
+		rq:                    rq,
+		logger:                logger,
+		sinceZeroValueCreator: sinceZeroValueCreator,
```
Comment (adamcfraser) [inline rest/blip_sync_messages.go]: A callback factory seems heavyweight here - can we just pass in a zeroed SequenceId that can be used as needed?

File: rest/blip_sync_messages.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+func (s *subChanges) since() db.SequenceID {
+
+	// Depending on the db sequence type, use correct zero sequence for since value
+	sinceSequenceId := s.sinceZeroValueCreator()
+
+	if sinceStr, found := s.rq.Properties["since"]; found {
+		var err error
+		if sinceSequenceId, err = db.ParseSequenceIDFromJSON([]byte(sinceStr)); err != nil {
+			s.logger.LogTo("Sync", "%s: Invalid sequence ID in 'since': %s", s.rq, sinceStr)
+			sinceSequenceId = db.SequenceID{}
```
Comment (adamcfraser) [inline rest/blip_sync_messages.go]: If this is trying to reset to zero, it would need to be using sinceZeroValueCreator().  In general it's a bit hard to follow the zero vs. value handling in subchanges.since() - might want to refactor to invoke sinceZeroValueCreator more intentionally instead of at the top of the function (e.g. when !found, or on error).

File: rest/blip_sync_messages.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	channelsArray := strings.Split(channelsParam, ",")
+	return channels.SetFromArray(channelsArray, channels.ExpandStar)
+}
+
+// Satisfy fmt.Stringer interface for dumping attributes of this subChanges request to logs
+func (s *subChanges) String() string {
+
+	buffer := bytes.NewBufferString("")
+
+	buffer.WriteString(fmt.Sprintf("Since:%v ", s.since()))
```
Comment (adamcfraser) [inline rest/blip_sync_messages.go]: s.since() looks like it recalculates since (and potentially logs errors on invalid since).  That doesn't feel like the right thing to be doing in String().

Comment (adamcfraser) [reply rest/blip_sync_messages.go]: More generally, this makes me wonder if we should just be building the logging string itself in handleSubChanges, instead of repeating all the request parameter parsing at log time.  

In particular, we're not trying to log a the state of mutable properties - we're just trying to log what the incoming request looked like.

File: rest/blip_sync_messages.go [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+		buffer.WriteString(fmt.Sprintf("Rev:%v ", rev))
+	}
+
+	return buffer.String()
+
+}
+
+type addRevision struct {
+	rq *blip.Message // The underlying BLIP message
+}
```
Comment (adamcfraser) [inline rest/blip_sync_messages.go]: As above - it feels like the helpers are in a bit of a middle ground, where they aren't quite doing enough to justify their existence.

If they were getting rid of some of the hard-to-parse code in the handlers themselves (like the `found && del != "0" && del != "false"` mentioned in an earlier comment), then I think there could be value.  

If parsing the request was computationally expensive, and the helpers were ensuring that the parsing only happened once, that might be useful.  

As they currently stand, they feel like they are just adding a layer of indirection that makes it harder to see what the calling code is doing with the request.

File: rest/blip_sync.go [hunk 6]
```diff
@@ -503,26 +526,28 @@ func (bh *blipHandler) sendRevision(sender *blip.Sender, seq db.SequenceID, docI
...
 	// Doc metadata comes from the BLIP message metadata, not magic document properties:
-	docID, found := rq.Properties["id"]
-	revID, rfound := rq.Properties["rev"]
+	docID, found := addRevision.id()
+	revID, rfound := addRevision.rev()
 	if !found || !rfound {
 		return base.HTTPErrorf(http.StatusBadRequest, "Missing docID or revID")
 	}
-	if del, found := rq.Properties["deleted"]; found && del != "0" && del != "false" {
+	if del, found := addRevision.deleted(); found && del != "0" && del != "false" {
```
Comment (adamcfraser) [inline rest/blip_sync.go]: I feel like we'd be making progress if addRevision.deleted() was getting rid of code like "found && del != "0" && del != "false"

**Review by adamcfraser — CHANGES_REQUESTED** [review]
A few minor comments, and then a larger concern about the overhead associated with the helper approach.

File: rest/blip_sync.go [hunk 7]
```diff
@@ -124,30 +132,36 @@ func (ctx *blipSyncContext) register(profile string, handlerFn func(*blipHandler
...
-		ctx.LogTo("Sync", "%s %q ... %s", rq, profile, ctx.effectiveUsername)
+		startTime := time.Now()
 
 		db, _ := db.GetDatabase(ctx.dbc, ctx.user)
 		handler := blipHandler{
 			blipSyncContext: ctx,
 			db:              db,
 		}
 
+		ctx.incrementSerialNumber()
```
[the remark this answers is not in the document]
Comment (tleyden) [reply rest/blip_sync.go]: I changed it so that the blipSyncContext holds the serial number counter, and whenever a new blipHandler is created it will increment the counter and store a snapshot of the counter in the blipHandler, which it will use for it's logging.

Example logs: https://gist.github.com/tleyden/8d75c0fba4cf5aaba025aad48bbbc92b

There are two unique blipSyncContexts:

* [2c49c010]
* [7b6b2a1f]

File: rest/blip_sync.go [hunk 8]
```diff
@@ -160,7 +174,11 @@ func (ctx *blipSyncContext) LogTo(key string, format string, args ...interface{}
 
 // Received a "getCheckpoint" request
 func (bh *blipHandler) handleGetCheckpoint(rq *blip.Message) error {
-	docID := "checkpoint/" + rq.Properties["client"]
+
+	client := rq.Properties["client"]
+	bh.logEndpointEntry(rq.Profile(), NewAdhocStringer(fmt.Sprintf("Client:%s", client)))
+
+	docID := fmt.Sprintf("checkpoint/%s", client)
```
[the remark this answers is not in the document]
Comment (tleyden) [reply rest/blip_sync.go]: > I think client is immutable for a given replication, isn't it?

I'm assuming so.

File: rest/blip_sync.go (hunk 6 above)
Comment (tleyden) [reply rest/blip_sync.go]: I moved that logic into `addRevision.deleted()` and added isolated test coverage: https://github.com/couchbase/sync_gateway/pull/3246/commits/367f92888ff98b5bb98c22dbafbc9048cda26eeb

That's the primary motivation of trying to refactor these helpers out -- to be able to have isolated unit tests as they increase in complexity or edge case bugs arise.

Comment (tleyden) [reply rest/blip_sync.go]: > That's the primary motivation of trying to refactor these helpers out -- to be able to have isolated unit tests as they increase in complexity or edge case bugs arise.

Actually I take that back .. the primary motivation was to be able to concisely emit the parameters of the request in single `String()` call.

File: rest/blip_sync_messages.go (hunk 2 above)
Comment (tleyden) [reply rest/blip_sync_messages.go]: Fixed in https://github.com/couchbase/sync_gateway/pull/3246/commits/5c6d570d7d852ccf297d7c3f1633149604ba629a

File: rest/blip_sync_messages.go (hunk 1 above)
Comment (tleyden) [reply rest/blip_sync_messages.go]: Fixed in https://github.com/couchbase/sync_gateway/issues/3249

File: rest/blip_sync_messages.go (hunk 3 above)
Comment (tleyden) [reply rest/blip_sync_messages.go]: Fixed in https://github.com/couchbase/sync_gateway/issues/3249

File: rest/blip_sync_messages.go (hunk 4 above)
Comment (tleyden) [reply rest/blip_sync_messages.go]: >  this makes me wonder if we should just be building the logging string itself in handleSubChanges

Currently for subChanges this is ~35 lines of code:

```
// Satisfy fmt.Stringer interface for dumping attributes of this subChanges request to logs
func (s *subChanges) String() string {

	buffer := bytes.NewBufferString("")

	since, err := s.since()
	if err != nil {
		base.Warn("Error discovering since value from subchanges.  Err: %v", err)
	} else {
		buffer.WriteString(fmt.Sprintf("Since:%v ", since))
	}

	continuous := s.continuous()
	if continuous {
		buffer.WriteString(fmt.Sprintf("Continuous:%v ", continuous))
	}

	activeOnly := s.activeOnly()
	if activeOnly {
		buffer.WriteString(fmt.Sprintf("ActiveOnly:%v ", activeOnly))
	}

	filter := s.filter()
	if len(filter) > 0 {
		buffer.WriteString(fmt.Sprintf("Filter:%v ", filter))
		channels, found := s.channels()
		if found {
			buffer.WriteString(fmt.Sprintf("Channels:%v ", channels))
		}
	}

	batchSize := s.batchSize()
	if batchSize != int(BlipDefaultBatchSize) {
		buffer.WriteString(fmt.Sprintf("BatchSize:%v ", s.batchSize()))
	}

	return buffer.String()

}
```

I guess it could be condensed, but it feels like it would have to be condensed quite a bit to fit in `handleSubChanges()` while keeping `handleSubChanges()` readable.  It could be a method rather than a helper struct, but the helper struct that conforms to the `fmt.Stringer()` interface allows some additional consistency / reuse in the code

**Comment by tleyden** [comment]
I pushed a new commit which fixed the drone ci build

File: rest/blip_sync.go [hunk 9]
```diff
@@ -171,7 +193,7 @@ func (bh *blipHandler) handleGetCheckpoint(rq *blip.Message) error {
 		return err
 	}
 	if value == nil {
-		return base.HTTPErrorf(401, "Not found")
+		return base.HTTPErrorf(404, "Not found")
```
Comment (adamcfraser) [inline rest/blip_sync.go]: Are we sure the client isn't expecting a 401 for a missing checkpoint?

Comment (adamcfraser) [reply rest/blip_sync.go]: Also, if we're cleaning this up, should we replace the '404' with http.StatusNotFound?

File: base/constants.go [hunk 10]
```diff
@@ -60,6 +60,7 @@ const (
 
 	DefaultViewQueryPageSize = 5000 // This must be greater than 1, or the code won't work due to windowing method
 
+	FilterByChannel = "sync_gateway/bychannel"
```
Comment (adamcfraser) [inline base/constants.go]: If we're going to make this a constant, we should update all the other existing usages (in changes_api, admin_api).

File: rest/blip_sync.go [hunk 11]
```diff
@@ -205,39 +231,44 @@ func (bh *blipHandler) handleSetCheckpoint(rq *blip.Message) error {
 // Received a "subChanges" subscription request
 func (bh *blipHandler) handleSubscribeToChanges(rq *blip.Message) error {
 
-	// Depeding on the db sequence type, use correct zero sequence for since value
-	since := bh.db.CreateZeroSinceValue()
+	subChanges := newSubChanges(rq, bh.blipSyncContext, bh.db.CreateZeroSinceValue(), bh.db.ParseSequenceID)
```
Comment (adamcfraser) [inline rest/blip_sync.go]: Are the new helpers really just request parameter helpers?  If so, could we rename to something like subChangesParams?  I think it would make their usage more transparent here.

File: rest/blip_sync.go [hunk 12]
```diff
@@ -665,17 +706,33 @@ func isCompressible(filename string, meta map[string]interface{}) bool {
 	return true // be optimistic by default
 }
 
+func (bh *blipHandler) logEndpointEntry(profile string, endpoint fmt.Stringer) {
+	bh.LogTo("SyncMsg", "#%03d: Prof:%s %s User:%s", bh.serialNumber, profile, endpoint, bh.effectiveUsername)
+}
```
Comment (adamcfraser) [inline rest/blip_sync.go]: What's the benefit in passing endpoint as a stringer, instead of just a string?  Given the choice between:

```
bh.logEndpointEntry(rq.Profile(), getAttachment)
bh.logEndpointEntry(rq.Profile(), NewAdhocStringer(fmt.Sprintf("#Changes: %d", len(changeList))))
```
versus
```
bh.logEndpointEntry(rq.Profile(), getAttachment.String())
bh.logEndpointEntry(rq.Profile(), fmt.Sprintf("#Changes: %d", len(changeList)))
```
...I think the latter is much easier to read/write.

File: rest/utilities_testing.go [hunk 13]
```diff
@@ -813,6 +813,7 @@ func EnableBlipSyncLogs() {
 	base.EnableLogKey("HTTP+")
 	base.EnableLogKey("Sync")
 	base.EnableLogKey("Sync+")
-	base.EnableLogKey("Sync++")
-
+	base.EnableLogKey("SyncMsg")
+	base.EnableLogKey("SyncMsg+")
+	base.EnableLogKey("WS")
```
Comment (adamcfraser) [inline rest/utilities_testing.go]: Might SyncMsg, SyncMsg+ and WS be too verbose to enable on every blip test?

File: rest/handler.go [hunk 14]
```diff
@@ -515,22 +515,27 @@ func (h *handler) getBearerToken() string {
...
 func (h *handler) currentEffectiveUserName() string {
 	var effectiveName string
 
 	if h.privs == adminPrivs {
-		effectiveName = " (as ADMIN)"
+		effectiveName = "ADMIN"
 	} else if h.user != nil {
 		if h.user.Name() != "" {
-			effectiveName = fmt.Sprintf(" (as %s)", h.user.Name())
+			effectiveName = fmt.Sprintf("%s", h.user.Name())
```
Comment (adamcfraser) [inline rest/handler.go]: Don't need Sprintf here any more.

**Review by adamcfraser — CHANGES_REQUESTED** [review]
I'm willing to live with the request helpers, but see the comment about potentially renaming them as parameter helpers.

A few other comments/requested changes, but no other concerns about the overall approach.

File: base/constants.go (hunk 10 above)
Comment (tleyden) [reply base/constants.go]: I'm just going to rollback this change, since there are so many places we could be using constants instead of strings, if we decide to do this, it probably makes more sense to do this comprehensively in a separate pass.

File: rest/blip_sync.go (hunk 9 above)
Comment (tleyden) [reply rest/blip_sync.go]: I filed https://github.com/couchbase/couchbase-lite-core/issues/384, and will roll it back out of this PR until we get confirmation

File: rest/blip_sync.go (hunk 11 above)
Comment (tleyden) [reply rest/blip_sync.go]: Yeah, that's basically what they are.  I renamed them all accordingly.

Comment (tleyden) [reply rest/blip_sync.go]: Another naming possibility would be:

- subChangesRequest
- revRequest 
- etc..

File: rest/blip_sync.go (hunk 12 above)
Comment (tleyden) [reply rest/blip_sync.go]: I think I had something in mind originally that never ended up being useful, so I'll switch this back to just plain strings.

File: rest/handler.go (hunk 14 above)
Comment (tleyden) [reply rest/handler.go]: Fixed

File: rest/utilities_testing.go (hunk 13 above)
Comment (tleyden) [reply rest/utilities_testing.go]: Ok I removed them.  I guess they can be added when needed for debugging.

**Comment by tleyden** [comment]
Addressed last round as a "review".

**Comment by tleyden** [comment]
Per https://github.com/couchbase/couchbase-lite-core/issues/384#issuecomment-362131531, changing from 401 -> 404 and using the the http lib status text.

Integration test:

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/341/

File: rest/blip_sync.go [hunk 15]
```diff
@@ -231,8 +231,8 @@ func (bh *blipHandler) handleSetCheckpoint(rq *blip.Message) error {
 // Received a "subChanges" subscription request
 func (bh *blipHandler) handleSubscribeToChanges(rq *blip.Message) error {
 
-	subChanges := newSubChanges(rq, bh.blipSyncContext, bh.db.CreateZeroSinceValue(), bh.db.ParseSequenceID)
-	bh.logEndpointEntry(rq.Profile(), subChanges)
+	subChanges := newSubChangesParams(rq, bh.blipSyncContext, bh.db.CreateZeroSinceValue(), bh.db.ParseSequenceID)
```
Comment (adamcfraser) [inline rest/blip_sync.go]: I feel like changing the local variable from subChanges to subChangesParams would actually be a big improvement in readability.   It makes subsequent usage (like subChanges.since()) much more self-documenting.  (i.e. I don't have to work out whether subChanges.since() refers to the current (mutable) since value in a continuous replication, or if it's just the request parameter).

Not critical to change if you think it makes the code too verbose.

File: rest/blip_sync.go [hunk 16]
```diff
@@ -665,17 +706,23 @@ func isCompressible(filename string, meta map[string]interface{}) bool {
 	return true // be optimistic by default
 }
 
+func (bh *blipHandler) logEndpointEntry(profile, endpoint string) {
+	// Prof is short for "Profile", which is basically the blip message type
+	bh.LogTo("SyncMsg", "#%03d: Prof:%s %s User:%s", bh.serialNumber, profile, endpoint, bh.effectiveUsername)
```
Comment (adamcfraser) [inline rest/blip_sync.go]: We've switched 'Prof' to 'Type' everywhere else, didn't we?

**Review by adamcfraser — CHANGES_REQUESTED** [review]
The variable naming is a style point - ignore if you disagree.

I think the logging item needs to be fixed, though.

File: rest/blip_sync.go (hunk 15 above)
Comment (tleyden) [reply rest/blip_sync.go]: Yeah, I missed that and will update.

File: rest/blip_sync.go (hunk 16 above)
Comment (tleyden) [reply rest/blip_sync.go]: Yep, thanks for catching

**Comment by tleyden** [comment]
Latest round of PR feedback addressed in https://github.com/couchbase/sync_gateway/pull/3246/commits/9509f5eff225743cd7ab73f08ed7d43dc3a62d4e + https://github.com/couchbase/sync_gateway/pull/3246/commits/21576ab26235da8eb1c0226fe6c3849626894757

**Review by adamcfraser — APPROVED** [review]
